sockeye/layers.py

```diff
@@ -46,11 +48,1542 @@ def activation(data: mx.sym.Symbol, act_type: str) -> mx.sym.Symbol:
         # Approximation of x * gaussian_cdf(x) used by Hendrycks and Gimpel
         return 0.5 * data * (1 + mx.sym.Activation((math.sqrt(2 / math.pi) * (data + (0.044715 * (data**3)))),
                                                    act_type="tanh"))
+    elif act_type == C.NO_ACTIVATION:
+        return data
     else:
         return mx.sym.Activation(data, act_type=act_type)
 
 
-class LayerNormalization:
+class Layer(ABC):
+
+    def get_max_seq_len(self) -> Optional[int]:
+        """
+        :return: The maximum length supported by the layer if such a restriction exists.
+        """
+        return None
+
+    @abstractmethod
+    def get_num_hidden(self) -> int:
+        """
+        :return: The representation size of this layer.
+        """
+        raise NotImplementedError()
+
+
+class EncoderLayer(Layer):
+    """
+    Generic encoder layer interface for a layer which takes the previous hidden state and sequence lengths and produces
+    a new hidden state, potentially changing the sequence length.
+    """
+
+    @abstractmethod
+    def encode_sequence(self,
+                        source_encoded: mx.sym.Symbol,
+                        source_encoded_lengths: mx.sym.Symbol,
+                        source_encoded_max_length: int,
+                        att_dict: Optional[Dict[str, mx.sym.Symbol]]) -> Tuple[mx.sym.Symbol, mx.sym.Symbol, int]:
+        """
+        Encodes data given sequence lengths of individual examples and maximum sequence length.
+
+        :param source_encoded: Input data of size (batch_size, seq_len, num_hidden).
+        :param source_encoded_lengths: Vector with sequence lengths of size (batch_size,).
+        :param source_encoded_max_length: Maximum sequence length.
+        :param att_dict: A dictionary of attention matrices used for visualization.
+            Each matrix must be of size (batch_size, source_length, source_length).
+        :return: Encoded versions of the input data, the new sequence lenghts and the new maximum length.
+        """
+        pass
+
+    def att_names(self) -> List[str]:
+        """Names of attention matrices produced by this layer."""
+        return []
+
+    def get_encoded_seq_len(self, seq_len: int) -> int:
+        """
+        :return: The size of the encoded sequence.
+        """
+        return seq_len
+
+
+class DecoderLayer(Layer):
+    """
+    Generic decoder layer interface.
```

Comment:
could benefit from a more detailed docstring describing the difference between decode_sequence and decode_step

Reply:
that is definitely true.